docs/contracts.md

```diff
@@ -0,0 +1,269 @@
+- its state can change. This can happen due to reentrancy.
+- its own balance can change. This can happen either because the contract is
+  transferring some of its tokens or, again, because of reentrancy.
+
+When an invocation fails none of these are changed, and any accrued changes are
+rolled back.
+
+If an invocation succeeds, then in the return value there is an indication on
+whether the state might have changed or not. "Might have changed" is currently
+defined as any of contract's entrypoints being executed successfully in the
+middle of handling the interrupt.
+
+TODO: The semantics of this will change with new state.
+
+There is no direct indication of whether the balance has changed. However
+**after** the invoke call, the host function `get_receive_self_balance` returns
+the **new** self balance.
+
+## Returning from a contract
+
+A contract can produce a return value, in addition to the return code. The
+return value is modelled as a writable buffer. If nothing is written, it is
+empty.
+
+```rust
+/// Write to the return value of the contract. The parameters are
+///
+/// - `start` the pointer to the location in memory where the data resides
+/// - `length` the size of data (in bytes)
+/// - `offset` where in the return value to write the data
+///
+/// The return value indicates how many bytes were written.
+fn write_output(start: *const u8, length: u32, offset: u32) -> u32;
+```
+
+## Parameters and observing return values from calls
+
+Each contract invocation starts with a parameter. If then a contract invokes
+another contract, which produces a return value, that return value is added to
+the stack of parameters the original contract can observe.
+
+```rust
+/// Get the size of the `i`-th parameter to the call. 0-th parameter is
+/// always the original parameter that the method was invoked with,
+/// calling invoke adds additional parameters to the stack. Returns `-1`
+/// if the given parameter does not exist.
+fn get_parameter_size(i: u32) -> i32;
+/// Write a section of the `i`-th parameter to the given location. Return
+/// the number of bytes written or `-1` if the parameter does not exist.
+/// The location is assumed to contain enough memory to
+/// write the requested length into.
+fn get_parameter_section(
+    i: u32,
+    param_bytes: *mut u8,
+    length: u32,
+    offset: u32,
+) -> i32;
+```
+
+## Obtaining information about the invoker
+
+Obtain policies on all credentials of the invoker. If the invoker is a contract
+it is the credentials of the owner that are returned. If it is an account, then
+the credentials of the account are returned.
+
+```rust
+/// Write a section of the policy to the given location. Return the number
+/// of bytes written. The location is assumed to contain enough memory to
+/// write the requested length into.
+fn get_policy_section(policy_bytes: *mut u8, length: u32, offset: u32) -> u32;
+```
+
+## Logging
+
+Contracts can output a small amount of data during execution. This data is
```

Comment:
How much is small?

Reply:
I added pointers to the constants. It's 512 * 64 bytes.